lib/chai/core/assertions.js

```diff
@@ -988,9 +988,28 @@ module.exports = function (chai, _) {
 
   function assertInstanceOf (constructor, msg) {
     if (msg) flag(this, 'message', msg);
+
+    var target = flag(this, 'object')
+    var validInstanceOfTarget = constructor === Object(constructor) && (
+        typeof constructor === 'function' ||
+        (typeof Symbol !== 'undefined' && typeof constructor[Symbol.hasInstance] === 'function')
+    );
+
+    if (!validInstanceOfTarget) {
+      var constructorType = constructor === null ? 'null' : typeof constructor;
+      throw new Error('The instanceof assertion needs a constructor but ' + constructorType + ' was given.');
+    }
+
+    var isInstanceOf;
+    try {
+        isInstanceOf = target instanceof constructor
+    } catch (err) {
+        throw new Error(target + ' instanceof ' + constructor + ' failed: ' + err.message + '.')
+    }
+
     var name = _.getName(constructor);
```

Comment:
Slight issue with error message wording in rare circumstances.

Example:

```js
var FakeConstructor = {};
FakeConstructor[Symbol.hasInstance] = function (val) {
  return val === 3;
};

var fakeInstanceB = 4;

expect(fakeInstanceB).to.be.an.instanceof(FakeConstructor);
```

Current output:

> AssertionError: expected 4 to be an instance of undefined

Output after https://github.com/chaijs/get-func-name/pull/20 is merged:

> AssertionError: expected 4 to be an instance of null

I'm thinking that if the result of `getName` is `null`, then the best thing we can do is just print out the `constructor` object; `null` will be meaningless/misleading to the end user. In this case, that'd be:

> AssertionError: expected 4 to be an instance of [object Object]

Not particularly helpful, but at least it's not misleading. An alternative would be to say "of an unnamed constructor".

> AssertionError: expected 4 to be an instance of an unnamed constructor